File: opencv-python/codeferm/CameraCalibration.py
"""
Copyright (c) Steven P. Goldsmith. All rights reserved.

Created by Steven P. Goldsmith on March 30, 2015
sgoldsmith@codeferm.com
"""

import logging, sys, os, time, cv2, numpy, argparse, glob, pickle

"""Camera calibration.

You need at least 10 images that pass cv2.findChessboardCorners at varying
angles and distances from the camera. You must do this for each resolution you
wish to calibrate. Camera matrix and distortion coefficients are pickled to
files for later use with undistort.

--inmask = Input file mask
--outdir = Output dir for debug images
--pattern = cols,rows of chess board

@author: sgoldsmith

"""

def splitFileName(fileName):
    """Split full path into file name components"""  
    path, fileName = os.path.split(fileName)
    name, ext = os.path.splitext(fileName)
    return path, name, ext

def saveArray(fileName, array):
    """Pickle ndarray"""  
    output = open(fileName, 'wb')
    pickle.dump(array, output)
    output.close()

def loadArray(fileName, array):
    """Load pickled ndarray"""
    pklFile = open(fileName, 'rb')
    array = pickle.load(pklFile)
    pklFile.close()
    return array

def findCorners(image, patternSize):
    """Find chess board corners"""  
    squareSize = 1.0
    patternPoints = numpy.zeros((numpy.prod(patternSize), 3), numpy.float32)
    patternPoints[:,:2] = numpy.indices(patternSize).T.reshape(-1, 2)
    patternPoints *= squareSize
    found, corners = cv2.findChessboardCorners(image, patternSize)
    return found, corners, patternPoints

def getPoints(inMask, outDir, patternSize):
    """Process all images matching inMask and output debug images to outDir"""
    imageNames = glob.glob(inMask)
    objPoints = []
    imgPoints = []
    # Set the criteria for the cornerSubPix algorithm
    term = (cv2.TERM_CRITERIA_EPS+cv2.TERM_CRITERIA_COUNT, 30, 0.1)
    passed = 0
    # Process all images
    for fileName in imageNames:
        image = cv2.imread(fileName, 0)
        found, corners, patternPoints = findCorners(image, patternSize)
        # Found corners
        if found:
            logger.info("Chessboard found in: %s" % fileName)
            cv2.cornerSubPix(image, corners, (5, 5), (-1, -1), term)
            vis = cv2.cvtColor(image, cv2.COLOR_GRAY2BGR)
            # Draw the corners
            cv2.drawChessboardCorners(vis, patternSize, corners, found)
            path, name, ext = splitFileName(fileName)
            # Write off marked up images
            cv2.imwrite('%s/%s_output.bmp' % (outdir, name), vis)
            # Add image and object points to lists for calibrateCamera
            imgPoints.append(corners.reshape(-1, 2))
            objPoints.append(patternPoints)
            passed += 1          
        else:
            logger.error("Chessboard not found in: %s" % fileName)
    logger.info("Images passed cv2.findChessboardCorners: %d" % passed)
    # We assume all images the same size, so we use last one
    h, w = image.shape[:2]    
    return h, w, objPoints, imgPoints

def undistort(image, cameraMatrix, distCoefs):
    """Undistort image"""  
    h,  w = image.shape[:2]
    newCameraMtx, roi = cv2.getOptimalNewCameraMatrix(cameraMatrix, distCoefs, (w, h), 1, (w, h))
    # Undistort
    dst = cv2.undistort(image, cameraMatrix, distCoefs, None, newCameraMtx)
    # Crop the image
    x, y, w, h = roi
    dst = dst[y:y + h, x:x + w]
    return dst

def reprojectionError(objPoints, imgPoints, rVecs, tVecs, cameraMatrix, distCoefs):
    """Re-projection error gives a good estimation of just how exact the found parameters are. This should be as close to zero as possible."""
    totalError = 0
    totalPoints = 0  
    for i in xrange(len(objPoints)):
        reprojectedPoints, _ = cv2.projectPoints(objPoints[i], rVecs[i], tVecs[i], cameraMatrix, distCoefs)
        reprojectedPoints = reprojectedPoints.reshape(-1,2)
        totalError += numpy.sum(numpy.abs(imgPoints[i] - reprojectedPoints) **2)
        totalPoints += len(objPoints[i])        
    return numpy.sqrt(totalError / totalPoints)

def undistortAll(inMask, outDir, cameraMatrix, distCoefs):
    """Process all images matching inMask and output debug images to outDir"""
    imageNames = glob.glob(inMask)
    # Process all images
    for fileName in imageNames:
        image = cv2.imread(fileName, 0)
        dst = undistort(image, cameraMatrix, distCoefs)
        path, name, ext = splitFileName(fileName)
        cv2.imwrite('%s/%s_undistort_output.bmp' % (outDir, name), dst)
    
if __name__ == '__main__':
    # Configure logger
    logger = logging.getLogger("CameraCalibration")
    logger.setLevel("INFO")
    formatter = logging.Formatter("%(asctime)s %(levelname)-8s %(module)s %(message)s")
    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(formatter)
    logger.addHandler(handler)
    parser = argparse.ArgumentParser(description='Calibrate camera using series of chessboards')
    # Add arguments
    parser.add_argument('-i', '--inmask', type=str, help='Mask used for input files', required=False)
    parser.add_argument('-o', '--outdir', type=str, help='Directory to send debug files', required=False)
    parser.add_argument('-p', '--pattern', type=str, help='Checcboard pattern cols,rows', required=False)
    # Array for all arguments passed to script
    args = parser.parse_args()
    inmask = args.inmask 
    outdir = args.outdir
    pattern = args.pattern
    # Handle defaults 
    if inmask == None:
        inmask = "../../resources/2015*.jpg"
    if outdir == None:
        outdir = "../../output/"
    if pattern == None:
        pattern = "7,5"
    logger.info("OpenCV %s" % cv2.__version__)
    logger.info("Input mask: %s" % inmask)
    logger.info("Output dir: %s" % outdir)
    patternSize = eval(pattern)
    start = time.time()
    h, w, objPoints, imgPoints = getPoints(inmask, outdir, patternSize)
    rms, cameraMatrix, distCoefs, rVecs, tVecs = cv2.calibrateCamera(objPoints, imgPoints, (w, h), None, None)
    error = reprojectionError(objPoints, imgPoints, rVecs, tVecs, cameraMatrix, distCoefs)
    logger.info("Mean reprojection error: %s" % error)
    undistortAll(inmask, outdir, cameraMatrix, distCoefs)
    # Pickle numpy arrays
    saveArray('%s/camera_matrix.pkl' % outdir, cameraMatrix)
    saveArray('%s/dist_coefs.pkl' % outdir, distCoefs)
    elapse = time.time() - start
    logger.info("RMS: %s" % rms)
    logger.info("Camera matrix: %s" % cameraMatrix)
    logger.info("Distortion coefficients: %s" % distCoefs.ravel())    
    logger.info("Elapse time: %4.2f seconds" % elapse)    

File: opencv-cpp/cameracalibration/src/CameraCalibration.cpp
/*
 * Copyright (c) Steven P. Goldsmith. All rights reserved.
 *
 * Created by Steven P. Goldsmith on November 6, 2015
 * sgoldsmith@codeferm.com
 */

#include <glob.h>
#include <iostream>
#include <sys/time.h>
#include <sstream>
#include <vector>
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

/**
 * Set the criteria for the cornerSubPix algorithm.
 */
static const TermCriteria CRITERIA = TermCriteria(
		TermCriteria::EPS + TermCriteria::COUNT, 30, 0.1);
/**
 * Return vector of sting containing list of files names matching pattern.
 */
vector<string> globVector(const string& pattern) {
	glob_t glob_result;
	glob(pattern.c_str(), GLOB_TILDE, NULL, &glob_result);
	vector<string> files;
	for (unsigned int i = 0; i < glob_result.gl_pathc; ++i) {
		files.push_back(string(glob_result.gl_pathv[i]));
	}
	globfree(&glob_result);
	return files;
}

/**
 * Find chess board corners.
 */
vector<Point2f> getCorners(Mat gray, Size pattern_size, Size win_size, Size zone_size) {
	vector<Point2f> corners;
	if (findChessboardCorners(gray, pattern_size, corners)) {
		cornerSubPix(gray, corners, win_size, zone_size, CRITERIA);
	}
	return corners;
}

/**
 * Point3f corners.
 */
vector<Point3f> getCorner3f(Size pattern_size) {
	vector<Point3f> corners3f(pattern_size.height * pattern_size.width);
	double squareSize = 50;
	int cnt = 0;
	for (int i = 0; i < pattern_size.height; ++i) {
		for (int j = 0; j < pattern_size.width; ++j, cnt++) {
			corners3f[cnt] = Point3f(j * squareSize, i * squareSize, 0.0);
		}
	}
	return corners3f;
}

/**
 * Process all images matching inMask and output debug images to outDir.
 */
void getPoints(string in_mask, string out_dir, Size pattern_size) {
	vector<Point3f> corners3f = getCorner3f(pattern_size);
	// Get list of files to process.
	vector<string> files = globVector(in_mask);
	vector<vector<Point3f> > object_points(files.size());
	vector<vector<Point2f> > image_points(files.size());
	vector<Mat> images(files.size());
	int passed = 0;
	for (size_t i = 0, max = files.size(); i != max; ++i) {
		Mat mat = imread(files[i], CV_LOAD_IMAGE_GRAYSCALE);
		vector<Point2f> corners;
		Size win_size = Size(5, 5);
		Size zone_size = Size(-1, -1);
		corners = getCorners(mat, pattern_size, win_size, zone_size);
		// Process only images that pass getCorners
		if (!corners.empty()) {
			cout << "Chessboard found in: " << files[i] << endl;
			Mat vis;
			// Convert to color for drawing
			cvtColor(mat, vis, COLOR_GRAY2BGR);
			drawChessboardCorners(vis, pattern_size, corners, true);
			// Get just file name from path
			string just_file_name = files[i].substr(
					files[i].find_last_of("/") + 1, files[i].length());
			// File name to write to
			string write_file_name = out_dir
					+ just_file_name.substr(0, just_file_name.find_last_of("."))
					+ "-cpp.bmp";
			imwrite(write_file_name, vis);
			object_points[i] = corners3f;
			image_points[i] = corners;
			images[i] = mat;
			passed++;
		} else {
			cout << "Chessboard not found in: " << files[i] << endl;
		}
	}
}

/**
 * Camera calibration.
 *
 * You need at least 10 images that pass cv2.findChessboardCorners at varying
 * angles and distances from the camera. You must do this for each resolution
 * you wish to calibrate. Camera matrix and distortion coefficients are written
 * to files for later use with undistort. This code is based on
 * http://computervisionandjava.blogspot.com/2013/10/camera-cailbration.html,
 * but follows Python code closely (hence the identical values returned).
 *
 * argv[1] = input file mask or will default to "../../resources/2015*.jpg" if no
 * args passed.
 *
 * argv[2] = output dir or will default to "../../output/" if no args passed.
 *
 * argv[3]] = cols,rows of chess board or will default to "7,5" if no args
 * passed.
 *
 * @author sgoldsmith
 * @version 1.0.0
 * @since 1.0.0
 */
int main(int argc, char *argv[]) {
	cout << CV_VERSION << endl;
	int return_val = 0;
	string in_mask;
	string out_dir;
	Size pattern_size;
	// Parse args
	if (argc == 4) {
		in_mask = argv[1];
		out_dir = argv[2];
		stringstream ss(argv[3]);
		vector<int> result(2);
		int i = 0;
		// Parse width and height "7,5" for example
		while (ss.good()) {
			string substr;
			getline(ss, substr, ',');
			result[i++] = atoi(substr.c_str());
		}
		pattern_size = Size(result[0], result[1]);
	} else {
		in_mask = "../../resources/2015*.jpg";
		out_dir = "../../output/";
		pattern_size = Size(7, 5);
	}
	cout << "Input mask: " << in_mask << endl;
	cout << "Output dir: " << out_dir << endl;
	cout << "Pattern size: " << pattern_size << endl;
	getPoints(in_mask, out_dir, pattern_size);
	return return_val;
}
